torch/testing/_internal/compilable_quantized_tensor.py

```diff
@@ -0,0 +1,366 @@
+            else layout_tensor.layout
+        )
+        kwargs["dtype"] = dtype  # type: ignore[assignment]
+        kwargs["requires_grad"] = False  # type: ignore[assignment]
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        layout_tensor: MyDTypeLayout,
+        shape: torch.Size,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        self.layout_tensor = layout_tensor
+
+    """__tensor_flatten__ and __tensor_unflatten__ are used to desugar the tensor into native Tensors/attributes and
+    reconstruct the tensor subclass instance from the desugared tensor and attributes, these are required to define
+    a Tensor subclass for torch.compile support
+    """
+
+    def __tensor_flatten__(self):  # type: ignore[no-untyped-def]
+        """
+        Given the class, returns the fields of the class as two lists
+        The first one contains any tensor fields such as int_data and scale as keys to a dictionary
+        The second one contains all other non tensor type fields as values of a list
+        """
+        return ["layout_tensor"], [self.shape, self.dtype]
+
+    @classmethod
+    def __tensor_unflatten__(  # type: ignore[no-untyped-def]
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        """
+        Given the flattened data from above, returns a class instance
+        tensor_data_dict contains the tensor fields of the class as a dictionary
+        tensor_attributes contains all other non tensor type fields
+        """
+        layout_tensor = tensor_data_dict["layout_tensor"]
+        shape, dtype = tensor_attributes
+        return cls(
+            layout_tensor,
+            shape if outer_size is None else outer_size,
+            dtype=dtype,
+        )
+
+    """classmethod that converts from a floating point Tensor (fp32/fp16/bf16) to the current dtype
+    """
+
+    @classmethod
+    def from_float(
+        cls,
+        input_float: torch.Tensor,
+        layout_type: LayoutType = PlainLayoutType(),
+    ) -> torch.Tensor:
+        dtype = torch.int16
+        scale = input_float.abs().max() / 255
+        int_data = (input_float / scale).to(torch.int8)
+        layout_tensor = PlainMyDTypeLayout(int_data, scale, PlainLayoutType())
+        return cls(layout_tensor, input_float.shape)
+
+    """[Optional] We can overwrite layout property of the Tensor to represent different packing formats
+    """
+
+    @property
+    def layout_type(self) -> LayoutType:
+        return self.layout_tensor.get_layout_type()
+
+    def dequantize(self, output_dtype=None):  # type: ignore[no-untyped-def]
+        """We can define a dequantize method to convert the quantized tensor to a floating point tensor"""
+        if output_dtype is None:
+            output_dtype = torch.get_default_dtype()
+        int_data, scale = self.layout_tensor.get_plain()
```

Comment:
is this subclass meant to be used with training / fine-tuning?

I would definitely prefer that the example subclass we have be trainable and work with autograd (in both eager compile), so we can really advertise that it works in all of the composability tests we care about.

(My suspicion is because `self.layout_tensor.get_plain()` doesn't really seem differentiable)

Reply:
not yet, we can add this later, when @andrewor14 added a training example in https://github.com/pytorch/ao/tree/main/tutorials/developer_api_guide

Reply:
I would like for the subclass that we put in core to start out with training support :)